blobstore/blobstore.smithy

```diff
@@ -0,0 +1,208 @@
+// blobstore.smithy
+// A service that stores objects (blobs) in named containers
+
+// Tell the code generator how to reference symbols defined in this namespace
+metadata package = [{ 
+    namespace: "org.wasmcloud.interface.blobstore", 
+    crate: "wasmcloud-interface-blobstore" 
+}]
+
+namespace org.wasmcloud.interface.blobstore
+
+use org.wasmcloud.model#wasmbus
+use org.wasmcloud.model#U32
+use org.wasmcloud.model#U64
+
+/// The BlobStore service, provider side
+@wasmbus(
+    contractId: "wasmcloud:blobstore",
+    providerReceive: true )
+service Blobstore {
+    version: "0.1",
+    operations: [ 
+        CreateContainer,
+        FindContainer,
+        GetContainerInfo,
+        ListContainers,
+        RemoveContainers,
+        ListObjects,
+        RemoveObjects,
+        StartUpload,
+        UploadChunk,
+        StartDownload,
+        ReceiveChunk,
```

Comment:
Shouldn't `ReceiveChunk` be part of an `actorReceive` service?

Reply:
I thought so att first,  but then I realised that since the actor knows the number of chunks, it can just as well call the provider and get the chunks as return value. It allowed for a much simpler provider implementation, at least for me who has not written a provider that calls an actor.

Reply:
I'm just trying to picture the use cases. If an actor has to ask for a set of chunks in a loop, that will block the actor for that entire period of time. However, if the actor is receiving chunks in a stream, then the actor can actually handle requests between chunk receipts.

Reply:
This is the code from the "old" file system provider... when the actor calls "start download" the provider then spawns a background thread and dispatches file chunks in a loop:
```

    fn start_download(
        &self,
        actor: &str,
        request: StreamRequest,
    ) -> Result<Vec<u8>, Box<dyn Error + Sync + Send>> {
        info!("Received request to start download : {:?}", request);
        let actor = actor.to_string();
        let bpath = Path::join(
            &Path::join(
                &self.rootdir.read().unwrap(),
                sanitize_id(&request.container),
            ),
            sanitize_id(&request.id),
        );
        let byte_size = &bpath.metadata()?.len();
        let bfile = std::fs::File::open(bpath)?;
        let chunk_size = if request.chunk_size == 0 {
            chunks::DEFAULT_CHUNK_SIZE
        } else {
            request.chunk_size as usize
        };
        let xfer = Transfer {
            blob_id: sanitize_id(&request.id),
            container: sanitize_id(&request.container),
            total_size: *byte_size,
            chunk_size: chunk_size as _,
            total_chunks: *byte_size / chunk_size as u64,
            context: request.context,
        };
        let iter = Chunks::new(bfile, chunk_size);
        let d = self.dispatcher.clone();
        std::thread::spawn(move || {
            iter.enumerate().for_each(|(i, chunk)| {
                dispatch_chunk(&xfer, &actor, i, d.clone(), chunk);
            });
        });

        Ok(vec![])
    }
```